tests/src/test/java/org/eclipse/hono/tests/IntegrationTestSupport.java

```diff
@@ -639,30 +636,58 @@ public boolean isGatewayModeSupported() {
     /**
      * Deletes all temporary objects from the Device Registry which
      * have been created during the last test execution.
+     * <p>
+     * <strong>Note:</strong> This wil consume the test context.
      *
      * @param ctx The vert.x context.
      */
     public void deleteObjects(final VertxTestContext ctx) {
 
-        if (!devicesToDelete.isEmpty()) {
-            devicesToDelete.forEach((tenantId, devices) -> {
-                final Checkpoint deviceDeletion = ctx.checkpoint(devices.size());
-                devices.forEach(deviceId -> {
-                    registry.deregisterDevice(tenantId, deviceId).onComplete(ok -> deviceDeletion.flag());
-                });
-                LOGGER.debug("deleted {} devices from tenant {}", devicesToDelete.size(), tenantId);
-            });
-            devicesToDelete.clear();
-        }
+        // first delete devices
```

Comment:
what's wrong with the code that already exists and has worked for months?

Reply:
It assumes that you can delete devices and tenants at the same time. Which you can only, because the current implementations fail to verify the tenant before doing device operations.

Devices and tenants get deleted in parallel. This lead to test failures, when a tenant got deleted before a device. The deletion of a devices would fail, but that would be ignored. Still, the device exists in the data store.

When the next test comes along, creates the same tenant, the device will again be there, and fail the test.